server/src/main/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkers.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.repositories.blobstore;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionRunnable;
+import org.elasticsearch.common.CheckedBiConsumer;
+import org.elasticsearch.common.UUIDs;
+import org.elasticsearch.repositories.SnapshotShardContext;
+
+import java.io.IOException;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.Executor;
+import java.util.concurrent.PriorityBlockingQueue;
+import java.util.function.Consumer;
+
+import static org.elasticsearch.index.snapshots.blobstore.BlobStoreIndexShardSnapshot.FileInfo;
+
+/**
+ * ShardSnapshotWorkers performs snapshotting tasks in the order dictated by the PriorityQueue of snapshot tasks.
+ * Each enqueued shard to snapshot results in one @{@link ShardSnapshotTask} and zero or more @{@link FileSnapshotTask}.
+ */
+public class ShardSnapshotWorkers {
+    private static final Logger logger = LogManager.getLogger(ShardSnapshotWorkers.class);
+
+    private final int maxWorkers;
+    private final Object mutex = new Object();
+    private final BlockingQueue<ShardSnapshotTask> snapshotTasks = new PriorityBlockingQueue<>();
+    private final Executor executor;
+    private final Consumer<SnapshotShardContext> shardSnapshotter;
+    private final CheckedBiConsumer<SnapshotShardContext, FileInfo, IOException> fileSnapshotter;
+    private volatile int workerCount = 0;
+
+    class ShardSnapshotTask implements Comparable<ShardSnapshotTask> {
+        protected final SnapshotShardContext context;
+
+        ShardSnapshotTask(SnapshotShardContext context) {
+            this.context = context;
+        }
+
+        public SnapshotShardContext context() {
+            return context;
+        }
+
+        public void run() {
+            shardSnapshotter.accept(context);
+        }
+
+        public short priority() {
+            return 1;
+        }
+
+        @Override
+        public int compareTo(ShardSnapshotTask other) {
+            int res = context.snapshotId().compareTo(other.context.snapshotId());
```

Comment:
There were no more replies to my question/suggestion [here](https://github.com/elastic/elasticsearch/pull/88209#discussion_r927741572), so I went for this as a first shot. Changing it would be easy once we agree on it. (I would need a bit more code since startTime is not part of the context).

So it should be ALL meta work (across all snapshots of the same repository) first, and then among file snapshots, sort by start time? No need to sort by snapshot name/id? Would start time be a replacement for sorting by snapshot name to prevent concurrent snapshots getting interleaved?

My understanding is that it is desirable to have predictable ordering here, and also finish shard snapshots as quick as possible. If a shard snapshot is unchanged, prioratizing it, and when it comes to file snapshots, finish the relevant ones (for the same snapshot) first, so that snapshots are marked as finished one at a time and do not get interleaved.